src/sync.rs

```diff
@@ -74,58 +75,137 @@ impl SyncSession {
         SA: ToSocketAddrs,
         T: AsRef<[u8]>,
     {
-        let socket = match destination.to_socket_addrs()?.next() {
-            Some(SocketAddr::V4(_)) => UdpSocket::bind((Ipv4Addr::new(0, 0, 0, 0), 0))?,
-            Some(SocketAddr::V6(_)) => UdpSocket::bind((Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0), 0))?,
-            None => panic!("empty list of socket addrs"),
-        };
+        let destination_out: SocketAddr = destination
+            .to_socket_addrs()?
+            .next()
+            .expect("empty list of socket addrs");
 
-        socket.set_read_timeout(timeout)?;
-        socket.connect(destination)?;
+        let socket = Self::create_socket(destination_out, timeout, false)?;
 
         let community = community
             .map(|c| c.as_ref().into())
             .unwrap_or_else(|| b"public".to_vec());
 
         Ok(SyncSession {
+            destination: destination_out,
             socket,
             community,
             req_id: Wrapping(starting_req_id),
             send_pdu: pdu::Buf::default(),
-            recv_buf: [0; 4096],
         })
     }
 
-    fn send_and_recv(socket: &UdpSocket, pdu: &pdu::Buf, out: &mut [u8]) -> SnmpResult<usize> {
-        if let Ok(_pdu_len) = socket.send(&pdu[..]) {
-            match socket.recv(out) {
-                Ok(len) => Ok(len),
-                Err(_) => Err(SnmpError::ReceiveError),
-            }
+    fn create_socket(
+        sock_addr: SocketAddr,
+        timeout: Option<Duration>,
+        broadcast: bool,
+    ) -> io::Result<UdpSocket> {
+        let socket = match sock_addr {
+            SocketAddr::V4(_) => UdpSocket::bind((Ipv4Addr::new(0, 0, 0, 0), 0)),
+            SocketAddr::V6(_) => UdpSocket::bind((Ipv6Addr::new(0, 0, 0, 0, 0, 0, 0, 0), 0)),
+        }?;
+        socket.set_read_timeout(timeout)?;
+        socket.set_broadcast(broadcast)?;
+        Ok(socket)
+    }
+
+    fn send_and_recv(&self, pdu: &pdu::Buf) -> SnmpResult<ResponseItemInt> {
+        if let Ok(_pdu_len) = self.socket.send_to(&pdu[..], self.destination) {
+            Self::recv_one(&self.socket)
         } else {
             Err(SnmpError::SendError)
         }
     }
 
+    fn recv_one(socket: &UdpSocket) -> SnmpResult<ResponseItemInt> {
+        let mut buf_out = vec![0u8; BUFFER_SIZE];
+
+        if let Ok((size, src_addr)) = socket.recv_from(&mut buf_out[..]) {
+            unsafe {
```

Comment:
No unsafes. Convert to Vec using `buf_out[..size].to_vec()` below.

Reply:
Please remove unsafe. Use either `buf_out[..size].to_vec()` or `buf_out.truncate(size)`.